### PR TITLE
Adding Tests for GGUFWriter Class

### PR DESCRIPTION
**Pull request by mendax0110**
### Description
This pull request adds comprehensive tests for the GGUFWriter class. Previously, there were no tests available for this class.

### Changes Made
- Introduced tests for the following methods:
  - `test_add_uint8`
  - `test_add_int8`
  - `test_add_uint16`
  - `test_write_tensors_to_file`
  - `test_add_description`
  - `test_add_uint32`
  - `test_add_float32`

### Purpose
The purpose of these tests is to ensure the reliability, correctness, and expected behavior of the GGUFWriter methods under different scenarios. 

### Benefits
- Enhances code quality by ensuring existing functionalities work as intended.
- Guards against potential regressions by establishing a testing suite for future changes.

### Additional Notes
- All tests have been successfully validated and pass without issues.
- The tests leverage temporary directories and context managers to maintain a clean testing environment.

### CODE REVIEW

**Review by cebtenzzre — CHANGES_REQUESTED** [review]
You'll have to reduce the amount of copy-pasting by using one or more shared functions for this to be mergeable. You could also consider testing more than one feature at a time, which provides less fine-grained failure information but has the same level of coverage.

**Comment by cebtenzzre** [comment]
The tests don't work. There's a warning:
```
E/usr/lib/python3.11/unittest/case.py:622: ResourceWarning: unclosed file <_io.BufferedWriter name='/tmp/tmpuro9bqgy/test4.gguf'>
  with outcome.testPartExecutor(self):
ResourceWarning: Enable tracemalloc to get the object allocation traceback
```
And an error:
```
======================================================================
ERROR: test_add_description (__main__.TestGGUFWriter.test_add_description)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jared/src/forks/llama.cpp/gguf-py/tests/test_gguf.py", line 58, in test_add_description
    gguf_writer.write_kv_data_to_file()
  File "/home/jared/src/forks/llama.cpp/gguf-py/gguf/gguf_writer.py", line 87, in write_kv_data_to_file
    raise ValueError(f'Expected output file to contain the header, got {self.state}')
ValueError: Expected output file to contain the header, got WriterState.EMPTY
```
For several tests.

**Comment by mendax0110** [comment]
> The tests don't work. There's a warning:
> 
> ```
> E/usr/lib/python3.11/unittest/case.py:622: ResourceWarning: unclosed file <_io.BufferedWriter name='/tmp/tmpuro9bqgy/test4.gguf'>
>   with outcome.testPartExecutor(self):
> ResourceWarning: Enable tracemalloc to get the object allocation traceback
> ```
> 
> And an error:
> 
> ```
> ======================================================================
> ERROR: test_add_description (__main__.TestGGUFWriter.test_add_description)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/jared/src/forks/llama.cpp/gguf-py/tests/test_gguf.py", line 58, in test_add_description
>     gguf_writer.write_kv_data_to_file()
>   File "/home/jared/src/forks/llama.cpp/gguf-py/gguf/gguf_writer.py", line 87, in write_kv_data_to_file
>     raise ValueError(f'Expected output file to contain the header, got {self.state}')
> ValueError: Expected output file to contain the header, got WriterState.EMPTY
> ```
> 
> For several tests.

Bild 06 12 23 um 21 46

I cannot reproduce the errors you got. Can you give me the steps to reproduce these errors?

**Comment by cebtenzzre** [comment]
You must call write_header_to_file before write_kv_data_to_file. I know because I wrote those exceptions. The only reason I can think it would pass is if you are testing an old version of the gguf module - the latest is 0.6.0.

File: gguf-py/tests/test_gguf.py [hunk 1]
```diff
@@ -1,7 +1,83 @@
...
 
 
-def test_write_gguf() -> None:
-    pass
+class TestGGUFWriter(unittest.TestCase):
+    def setUp(self):
+        self.temp_dir = tempfile.TemporaryDirectory()
+
+    def tearDown(self):
+        self.temp_dir.cleanup()
```
Comment (ggerganov) [inline gguf-py/tests/test_gguf.py:15]: Should these functions be `snake_case`?

**Review by ggerganov — APPROVED** [review]

File: gguf-py/tests/test_gguf.py (hunk 1 above)
Comment (cebtenzzre) [reply gguf-py/tests/test_gguf.py:15]: They override unittest metods, and unittest is so old that it predates the standardization on snake_case.

**Comment by cebtenzzre** [comment]
This PR still doesn't work, please cd to gguf-py and run `pip install -e .` to test with the latest gguf package.

**Comment by mendax0110** [comment]
I did this. I will uninstall/install gguf again and try it again.

**Pull request closed without merging**